### PR TITLE
feat(map_loader): add grid coordinates for partial/differential map load

### PR DESCRIPTION
**Pull request by kminoda**
## Description
Please see https://github.com/orgs/autowarefoundation/discussions/3328 for the detail

You can use https://github.com/MapIV/pointcloud_divider for dividing pointcloud map into grids

Please find a how-to in [updated readme](https://github.com/kminoda/autoware.universe/tree/feat/add_grid_coordinates_in_map_msg/map/map_loader)
image

Should be merged after:
- https://github.com/autowarefoundation/autoware_msgs/pull/52

## Tests performed
- Generate a divided pointcloud map with a given tool
- Load the pointcloud map and its metadata with logging_simulator

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kminoda** [comment]
@RyuYamamoto Sorry again, this is the latest PR :pray:

File: map/map_loader/src/pointcloud_map_loader/pointcloud_map_loader_node.cpp [hunk 1]
```diff
@@ -66,18 +67,53 @@ PointCloudMapLoaderNode::PointCloudMapLoaderNode(const rclcpp::NodeOptions & opt
...
+    // Autoware users get used to handling the PCD file(s) with metadata.
+    RCLCPP_INFO_STREAM(get_logger(), "Create PCD metadata, as the pointcloud is a single file.");
+    pcl::PointCloud<pcl::PointXYZ> single_pcd;
+    if (pcl::io::loadPCDFile(pcd_paths[0], single_pcd) == -1) {
+      throw std::runtime_error("PCD load failed: " + pcd_paths[0]);
+    }
+    PCDFileMetadata metadata = {};
+    pcl::getMinMax3D(single_pcd, metadata.min, metadata.max);
+    pcd_metadata_dict[pcd_paths[0]] = metadata;
+    std::cout << "pointcloud_map_loader " << pcd_paths[0] << std::endl;
```
Comment (RyuYamamoto) [inline map/map_loader/src/pointcloud_map_loader/pointcloud_map_loader_node.cpp:114]: RCLCPP_INFO_STREAM?

Comment (kminoda) [reply map/map_loader/src/pointcloud_map_loader/pointcloud_map_loader_node.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/3205/commits/76c3bbd36d52f7e74a14d2a155af137a0ed3859d

**Review by RyuYamamoto — APPROVED** [review]
LGTM

**Comment by kminoda** [comment]
@yukkysaito @YamatoAndo Could you also briefly review this PR?
I especially would like you to verify that the directory structure looks OK to you.

### How to store map-related files

If you only have one pointcloud map, Autoware will assume the following directory structure by default.

```bash
sample-map-rosbag
├── lanelet2_map.osm
├── pointcloud_map.pcd
└── pointcloud_map_metadata.yaml (not necessary when only a single PCD file exists)
```

If you have multiple rosbags, an example directory structure would be as follows:

```bash
sample-map-rosbag
├── lanelet2_map.osm
├── pointcloud_map
│   ├── pcd_00.pcd
│   ├── pcd_01.pcd
│   └── pcd_02.pcd
└── pointcloud_map_metadata.yaml
```

Note that in this case, you have to launch autoware or simulators with a following command:

```bash
ros2 launch autoware_launch autoware.launch.xml map_path:=/PATH/TO/sample-map-rosbag pointcloud_map_file:=pointcloud_map ...
```

File: map/map_loader/README.md [hunk 2]
```diff
@@ -38,20 +38,70 @@ Please see [the description of `GetDifferentialPointCloudMap.srv`](https://githu
...
+│   ├── pcd_00.pcd
+│   ├── pcd_01.pcd
+│   └── pcd_02.pcd
+└── pointcloud_map_metadata.yaml
+```
+
+Note that in this case, you have to launch autoware or simulators with a following command:
+
+```bash
+ros2 launch autoware_launch autoware.launch.xml map_path:=/PATH/TO/sample-map-rosbag pointcloud_map_file:=pointcloud_map ...
```
Comment (yukkysaito) [inline map/map_loader/README.md:103]:
```suggestion
ros2 launch autoware_launch autoware.launch.xml map_path:=/PATH/TO/sample-map-rosbag pointcloud_map_file:=pointcloud_map ...
```
what meaning `...`? (continuing?)

Comment (kminoda) [reply map/map_loader/README.md:103]: I meant `...` = `vehicle_model:=VEHICLE_MODEL sensor_model:=SENSOR_MODEL other_args:=...`

Comment (kminoda) [reply map/map_loader/README.md]: Updated: https://github.com/autowarefoundation/autoware.universe/pull/3205/commits/b0ae10b923fd1375af2890226f292e6612ddd3a3

Comment (kminoda) [reply map/map_loader/README.md]: This command is just an example command when the map directory structure is as described in the readme. So in that sense it is not necessary.
```bash
sample-map-rosbag
├── lanelet2_map.osm
├── pointcloud_map
│   ├── pcd_00.pcd
│   ├── pcd_01.pcd
│   └── pcd_02.pcd
└── pointcloud_map_metadata.yaml
```

Comment (yukkysaito) [reply map/map_loader/README.md]: Okay. I don't want to increase the commandsif possible. :+1:

Comment (kminoda) [reply map/map_loader/README.md]: I just realized that this solves everything: https://github.com/autowarefoundation/autoware.universe/pull/3205/commits/271c3872d9d99959660ba14e9e99fbf70f533528
Resolved :+1:

**Pull request merged**